libbeat/common/transport: fix log message about TLS

belimawr opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

This commit fixes the log message issued by the `test output` command.
Our current TLS verification relies on more than the value of
`tlsConfig.InsecureSkipVerify`, so the previous implementation would
log that TLS was disabled when it was not.

This commit fixes it by checking the value of `config.Verification`.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

It fix a misleading log message

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [ ] ~~I have commented my code, particularly in hard-to-understand areas~~
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
- [ ]
-->

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

1. Configure any output with TLS enabled and `ssl.verification_mode` != `strict` (it always worked for `strict`).
2. Run the `test output` command from your beat
3. You should see the log: `server's certificate chain verification is enabled`
## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->

~~## Use cases~~

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

~~## Screenshots~~

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

~~## Logs~~

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane  (Team:Elastic-Agent-Data-Plane)

cmacknz commented:
I followed the code and it looks like the the `InsecureSkipVerify` mode is only set in one place based on the `Verification` mode. It looks there is another log message when using `VerifyNone` there.


https://github.com/elastic/beats/blob/3d3452f1cb00944cd965b9e7fc91dda4ecc61913/libbeat/common/transport/tlscommon/tls_config.go#L103-L114

We consider anything other than `VerifyStrict` as insecure and set `InsecureSkipVerify`.

`InsecureSkipVerify` says this: 

```go
	// InsecureSkipVerify controls whether a client verifies the server's
	// certificate chain and host name. If InsecureSkipVerify is true, crypto/tls
	// accepts any certificate presented by the server and any host name in that
	// certificate. In this mode, TLS is susceptible to machine-in-the-middle
	// attacks unless custom verification is used. This should be used only for
	// testing or in combination with VerifyConnection or VerifyPeerCertificate.
	InsecureSkipVerify bool
```

This change stops logging when this flag is set, which doesn't feel right based on what I'm reading. Why would we want to stop logging when `InsecureSkipVerify` is set? Should we reword the existing log message instead of changing when it is logged?

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30063/detail/PR-30063/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30063/detail/PR-30063/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30063/detail/PR-30063/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30063/detail/PR-30063/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-01-27T14:50:47.263Z&rangeTo=2022-01-27T15:10:47.263Z&transactionName=null&transactionType=job&latencyAggregationType=avg&traceId=471cacba75a201ebb08e65e2c69a3951&transactionId=a73e2f834aa3d307)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `Aborted from #2`


* Start Time: 2022-01-27T15:00:47.263+0000


* Duration: 122 min 34 sec


* Commit: 0050c544fc2d29fbc84b20f23ff439338478fd8c



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3961  |
| Skipped      | 370 |
| Total        | 4331   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->











### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30063/detail/PR-30063/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 2 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30063/runs/1/steps/8768/log/?start=0">here</a></li>
<li>Description: <code>untar: step failed with error null</code></l1>
</ul>
  
  </p>
  </details>


## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

belimawr commented on libbeat/common/transport/tls.go:
```diff
@@ -154,7 +154,7 @@ func tlsDialWith(
 	}
 
-	if tlsConfig.InsecureSkipVerify {
+	if config.Verification == tlscommon.VerifyNone {
```
A few lines below, in the same function you quoted we set the `tls.Config.VerifyConnection`: https://github.com/elastic/beats/blob/3d3452f1cb00944cd965b9e7fc91dda4ecc61913/libbeat/common/transport/tlscommon/tls_config.go#L120

From the Go docs about this function:
> 	// If normal verification fails then the handshake will abort before
	// considering this callback. This callback will run for all connections
	// regardless of InsecureSkipVerify or ClientAuth settings.
	VerifyConnection func(ConnectionState) error


This is the function that does all the TLS validation to `VerifyFull` and `VerifyCertificate`, for `VerifyStrict` this function only validates the pin as the rest is taken care by the standard library.

We do it like that because the normal TLS validation from Go's standard library (`InsecureSkipVerify = false`) does not allow for certificates with empty SNA, while `VerifyFull` and `VerifyCertificate` do.

The "logging bug" happens because [`tlsDialWith`](https://github.com/elastic/beats/blob/0250eef196a054c3c5efd36a946ae5017d0c271d/libbeat/common/transport/tls.go#L133) does not know about the interaction between  `InsecureSkipVerify` and our `VerifyConnection` function.

cmacknz replied:
Ah I see now, thanks. We are manually doing verification for the modes between None and Strict.

Add a comment in the code about this and I'm fine with this change.

kvch commented on CHANGELOG.next.asciidoc:
```diff
@@ -91,7 +91,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 
 *Affecting all Beats*
-
+- common/transport: Fix TLS status message from `test output` command. {pull}30063[30063]
```
Please do not include the package names in the changelog entry. This changelog is written for our users, they should not know about where the log message is located in Beats. I really like how you put it in the PR description, I would like you to explain the change similarly. Example: When it `ssl.verification_mode` was set to `full` or `certificate`, the command `test output` logged incorrectly that TLS was disabled.

kvch reviewed (commented):
Once you address the comments, I will approve.

belimawr commented:
@kvch @cmacknz I've addressed your comments on 627303874d.

@cmacknz I was quite detailed on the comment, let me know if it was too much :sweat_smile:

cmacknz commented on libbeat/common/transport/tls.go:
```diff
@@ -154,7 +154,14 @@ func tlsDialWith(
 	}
 
-	if tlsConfig.InsecureSkipVerify {
+	// We only chech the status of config.Verification (`ssl.verirication_mode`
```
```suggestion
	// We only check the status of config.Verification (`ssl.verification_mode`
```

belimawr replied:
Good catch!

cmacknz reviewed (approved):
Thanks, great comment.

belimawr commented:
@ruflin should we backport?

ruflin reviewed (approved):
LGTM, lets backport.
